### ISSUE TITLE
Include payload in term advisories

### ISSUE THREAD

**Issue by ripienaar** [nats-io/nats-server#5049]
### Proposed change

Termination advisories include all the message data to help one find the message in the stream that holds the message, but:

1. Streams can be removing messages from due to limits meaning the ability to fetch the data might be short lived
2. It's expensive to go feetch the message to build something like a DLQ

The proposal is to include all the original message data in the term advisory - original timestamp in addition to the timestamp the event was create and most importantly the origianal message payload.

### Use case

One can read these advisories and build a full DLQ that holds payload and all without aditional API integrations.

**Comment by amitamit281** [comment]
Hi @ripienaar 
Will the payload be published on the same existing subject or on a different subject, such as "JS.EVENT.ADVISORY.CONSUMER.MSG_TERMINATED.<stream name>.<consumer name>"? This inquiry stems from our concern regarding the presence of compressed or confidential data in the stream. We currently subscribe to these advisories and forward them to Datadog and Splunk for alerts. However, we wish to refrain from sending data to Splunk and Datadog. If possible, it would be preferable to publish the payload on a separate subject. This would allow us to subscribe payload to a Dead Letter Queue (DLQ) Stream and republish the data if necessary.

**Comment by ripienaar** [comment]
@amitamit281 can you not remove this field in the tool you use to integrate with Splunk? We would rather not go crazy with too many options and settings so trying to keep it easy

**Comment by amitamit281** [comment]
@ripienaar yes we can remove the payload. However, we also intend to store the payload in a stream. If it's published on a separate subject, it would be beneficial.

**Comment by Jarema** [comment]
@ripienaar For completness we should also include headers.

**Comment by bruth** [comment]
@ripienaar @Jarema would this be reasonable for the 2.11 timeline (seems straightforwarrd)? If so, feel free to set the milestone and assign someone.

**Comment by ripienaar** [comment]
@bruth no we decided to park it while we design a better DLQ solution due to comments about data leakage

**Comment by saty9** [comment]
Hi,

Is there still an intent to improve this? We are looking at implementing dead lettering for a new system we are trying to build with NATs. But the current state means we will need to keep a replica of all messages so we can retrieve dead ones (we are using a workqueue policy) which isnt a great situation and still vulnerable to the  issue mentioned above of limits will eventually remove the message too.

**Comment by ripienaar** [comment]
We do plan to improve DLQ processing but we have not spent focussed time on this since this issue - we know just blanket publishing the data wouldn't help or work for many

**Comment by saty9** [comment]
Ok is there any timeframe on that as this is a commonly used feature in rabbitmq that we were in the process of moving from onto nats and it would work for us and presumably other people who want dead letter queues a 2nd subject that can be subscribed to with this info or an option the server could be started with would let people opt into the functionality?

**Comment by ripienaar** [comment]
Nothing on immediate roadmap.  Unlike rabbitmq where a message basically is only ever going to be delivered once (successfully) stream based systems can deliver the same message many times to many different consumers for different purposes.  So DLQ is quite a complex subject, you need a DLQ per consumer not per stream.  And consumers can be ephemeral, short lived, single use etc.  What the right thing to do is going to be very dependent on your needs.

So we designed the system so you can solve it for your needs - you can consume the ACK events and fetch the message in code for reprocessing and WQs will not delete messages that were not successfully processed.

Here I have the advisories that some messages on a WQ reached max delivery attempts:

```
16:07:34 [JS Max Deliveries] WQ (1) > PULL: 2 deliveries
16:07:34 [JS Max Deliveries] WQ (2) > PULL: 2 deliveries
16:07:34 [JS Max Deliveries] WQ (3) > PULL: 2 deliveries
```

And those messages still exist.

```
[rip@s1-lon]% nats s get WQ 1
Item: WQ#1 received 2024-09-19 20:05:23.725407596 +0000 UTC (234d20h3m9s) on Subject js.in.wq.1

1
```

So if you process the advisories in real time you can easily do whatever you need and not have to keep all messages for all time - just set your limits in line with your operational timelines.

**Comment by saty9** [comment]
ok thats good to know. To clarify if you dont subscribe to advisories and use them to clean up messages in work queue streams will the streams slowly fill with dead messages? how do you detect and clear those out?

**Comment by ripienaar** [comment]
You set a max age limit on the stream.

**Comment by Axl116** [comment]
@ripienaar I'm trying to create a DLQ service that consumes the advisory subjects (MSG_TERMINATED, MAX_DELIVERIES) and retrieves the original message by stream sequence.

I'm using a WorkQueue stream and it would seem like when terminated, the msg is also deleted from the stream before I can retrieve it. 
You mention in one of your comments that "WQs will not delete messages that were not successfully processed", surely terminating a message should count as "not successfully processed"?

I'm on NATS Server version 2.10.25, is this too old? I have not started digging into the release notes yet.

**Comment by neilalexander** [comment]
2.10.25 is very old, yes, we're on 2.14.3 now. Please let us know if you still have problems on the latest version.

**Comment by Axl116** [comment]
@neilalexander Tested with 2.14.3 now. 

I have a .NET 8 worker service that attempts to send a request to an external endpoint, it receives an unrecoverable error and therefore calls `msg.AckTerminateAsync();`. This results in a `MSG_TERMINATED` advisory as expected, however the message is not retrievable. 

Using NATS CLI:
```
[#44] Received on "$JS.EVENT.ADVISORY.CONSUMER.MSG_TERMINATED.mystream.myconsumer"
{"type":"io.nats.jetstream.advisory.v1.terminated","id":"Xf2A0lr2Oc322EHlj3yWoB","timestamp":"2026-07-01T09:18:00.8148287Z","stream":"mystream","consumer":"myconsumer","consumer_seq":63,"stream_seq":390,"deliveries":1}
```

Then attempting to retrieve it using the `stream_seq`

```
C:\Users\***>nats stream get mystream 390
nats: error: could not retrieve mystream#390: no message found (10037)
```

Stream details:
```
Options:

        Retention: WorkQueue
       Acknowledgments: true
        Discard Policy: Old
      Duplicate Window: 2m0s
     Allows Msg Delete: true
          Allows Purge: true
        Allows Rollups: false

Limits:

      Maximum Messages: unlimited
   Maximum Per Subject: unlimited
         Maximum Bytes: unlimited
           Maximum Age: unlimited
  Maximum Message Size: unlimited
     Maximum Consumers: unlimited

Metadata:

           _nats.level: 4
             _nats.ver: 2.14.3
```

I also noticed that within the stream state details, `Deleted Messages` is incremented by one for each `MSG_TERMINATED` advisory.

**Comment by Axl116** [comment]
bump @neilalexander @ripienaar, thoughts on my question above?

**Comment by ripienaar** [comment]
This is unfortunately unavoidable, that's just how work queues work - Term'd messages are not kept
